dbus/methods_test.go

```diff
@@ -130,14 +135,209 @@ func TestStartStopUnit(t *testing.T) {
 	<-reschan
 
 	units, err = conn.ListUnits()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	unit = getUnitStatus(units, target)
+
+	if unit != nil {
+		t.Fatalf("Test unit found in list, should be stopped")
+	}
+}
+
+// Ensure that basic unit restarting works.
+func TestRestartUnit(t *testing.T) {
+	target := "start-stop.service"
+	conn := setupConn(t)
+
+	setupUnit(target, conn, t)
+	linkUnit(target, conn, t)
+
+	// Start the unit
+	reschan := make(chan string)
+	_, err := conn.StartUnit(target, "replace", reschan)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	job := <-reschan
+	if job != "done" {
+		t.Fatal("Job is not done:", job)
+	}
+
+	units, err := conn.ListUnits()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	unit := getUnitStatus(units, target)
+	if unit == nil {
+		t.Fatalf("Test unit not found in list")
+	} else if unit.ActiveState != "active" {
+		t.Fatalf("Test unit not active")
+	}
+
+	// Restart the unit
+	reschan = make(chan string)
+	_, err = conn.RestartUnit(target, "replace", reschan)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	job = <-reschan
+	if job != "done" {
+		t.Fatal("Job is not done:", job)
+	}
+
+	// Stop the unit
+	_, err = conn.StopUnit(target, "replace", reschan)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// wait for StopUnit job to complete
+	<-reschan
+
+	units, err = conn.ListUnits()
+	if err != nil {
+		t.Fatal(err)
+	}
 
 	unit = getUnitStatus(units, target)
+	if unit != nil {
+		t.Fatalf("Test unit found in list, should be stopped")
+	}
+
+	// Try to restart the unit.
+	// It should still succeed, even if the unit is inactive.
+	reschan = make(chan string)
+	_, err = conn.TryRestartUnit(target, "replace", reschan)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// wait for StopUnit job to complete
+	<-reschan
 
+	units, err = conn.ListUnits()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	unit = getUnitStatus(units, target)
 	if unit != nil {
 		t.Fatalf("Test unit found in list, should be stopped")
 	}
 }
 
+// Ensure that basic unit reloading works.
+func TestReloadUnit(t *testing.T) {
+	target := "reload.service"
+	conn := setupConn(t)
+
+	setupUnit(target, conn, t)
+	linkUnit(target, conn, t)
+
+	// Start the unit
+	reschan := make(chan string)
+	_, err := conn.StartUnit(target, "replace", reschan)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	job := <-reschan
+	if job != "done" {
+		t.Fatal("Job is not done:", job)
+	}
+
+	units, err := conn.ListUnits()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	unit := getUnitStatus(units, target)
+	if unit == nil {
+		t.Fatalf("Test unit not found in list")
+	} else if unit.ActiveState != "active" {
+		t.Fatalf("Test unit not active")
+	}
+
+	// Reload the unit
+	reschan = make(chan string)
+
+	// NOTE: As ReloadUnit() sometimes returns "failed", let's
+	// retry 3 times until it's returning the correct status.
```

Comment:
@lucab I've seen the race only on Fedora 24 with systemd v229. OTOH it seems to work fine on Debian/Ubuntu. The race doesn't always happen. Not sure if it's an upstream issue or a go-systemd issue.